src/dask_awkward/lib/optimize.py

```diff
@@ -97,166 +59,44 @@ def _projectable_io_layer_names(dsk: HighLevelGraph) -> list[str]:
-    # final necessary key is the 0th partition of the last layer in
-    # the graph (hence the toposort to find last layer).
-    final_key = (dsk._toposort_layers()[-1], 0)
-    new_hlg = HighLevelGraph(layers, deps).cull([final_key])
-    get_sync(new_hlg, list(new_hlg.keys()))
-
-
-def _necessary_columns_brute(dsk: HighLevelGraph) -> dict:
-    # staring fields should be those belonging to the AwkwardIOLayer's
-    # metadata (typetracer) array.
-    out_meta = list(dsk.layers.values())[-1]._meta  # type: ignore
-    keep = out_meta.layout.form.columns()
-    columns: list[str] = []
-
-    pio_layer_names = _projectable_io_layer_names(dsk)
-    if len(pio_layer_names) > 1:
-        raise RuntimeError(
-            "'brute' method of optimization currently only graphs with a single IO layer."
-        )
-    pio_layer = pio_layer_names[0]
-
-    columns = dsk.layers[pio_layer]._meta.layout.form.columns()  # type: ignore
-
-    # can only select output columns that exist in the input
-    # (other names may have come from aliases)
-    keep = [c for c in out_meta.layout.form.columns() if c in columns]
-
-    for c in columns:
-        if c in keep:
-            continue
-        holdout = c
-        allcolumns = set(columns)
-        remaining = list(allcolumns - {holdout})
-        try:
-            _attempt_compute_with_columns_brute(dsk, columns=remaining)
-        except IndexError:
-            keep.append(holdout)
-    if keep == columns:
-        keep = None
-    return {pio_layer: keep}
-
-
-def _has_projectable_awkward_io_layer(dsk: HighLevelGraph) -> bool:
-    for _, v in dsk.layers.items():
-        if isinstance(v, AwkwardIOLayer) and hasattr(v.io_func, "project_columns"):
-            return True
-    return False
-
-
-def optimize_iolayer_columns_brute(dsk: HighLevelGraph) -> HighLevelGraph:
-    # if the task graph doesn't contain a column-projectable
-    # AwkwardIOLayer then bail on this optimization (just return the
-    # existing task graph).
-    if not _has_projectable_awkward_io_layer(dsk):
-        return dsk
-    # determine the necessary columns to complete the executation of
-    # the metadata (typetracer) based task graph.
-    necessary_cols = _necessary_columns_brute(dsk)
-    layer_name, necessary_cols = list(_necessary_columns_brute(dsk).items())[0]
-
-    if necessary_cols is None:
-        return dsk  # type: ignore
-    layers = dsk.layers.copy()  # type: ignore
-    deps = dsk.dependencies.copy()  # type: ignore
-    new_layer: Layer | None = None
-
-    new_layer = dsk.layers[layer_name].project_columns(necessary_cols)  # type: ignore
-    layers[layer_name] = new_layer
-
+    deps = dsk.dependencies.copy()
+    for name, ttr in reports.items():
+        cols = set(ttr.data_touched)
+        select = []
+        for col in cols:
+            if col is None:
+                continue
+            n, c = col.split(".", 1)
+            if n == name:
```

Comment:
I think this is always True, at least for the simple linear graphs we have so far